powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/AssertionChallengeConverter.java

```diff
@@ -48,6 +55,65 @@ public AssertionChallengeResponse fromChallenge(AssertionChallenge source) {
         destination.setChallenge(source.getChallenge());
         destination.setFailedAttempts(source.getFailedAttempts());
         destination.setMaxFailedAttempts(source.getMaxFailedAttempts());
+        destination.setAllowCredentials(source.getAllowCredentials());
+        return destination;
+    }
+
+    /**
+     * Convert the assertion challenge request to a new operation create request. Optionally, store allowed
+     * authenticators with the operation.
+     *
+     * @param source Assertion challenge.
+     * @param authenticatorDetails Allowed authenticators. If null or empty, all are allowed.
+     * @return Request for creating a new operation.
+     */
+    public OperationCreateRequest convertAssertionRequestToOperationRequest(AssertionChallengeRequest source, List<AuthenticatorDetail> authenticatorDetails) {
+        final OperationCreateRequest destination = new OperationCreateRequest();
+        destination.setUserId(source.getUserId());
+        destination.setApplications(source.getApplicationIds());
+        destination.setTemplateName(source.getTemplateName());
+        destination.getParameters().putAll(source.getParameters());
+
+        if (authenticatorDetails != null && !authenticatorDetails.isEmpty()) {
+            final Set<String> allowCredentials = new LinkedHashSet<>();
+            for (AuthenticatorDetail ad : authenticatorDetails) {
+                allowCredentials.add(ad.getExternalId());
+            }
+            destination.setAdditionalData(Map.of("allowCredentials", allowCredentials));
```

Comment:
If possible do not put this logic - really on the `activationId`. But at least put proper externalized constant, i.e., ATTR_ALLOW_CREDENTIALS -  with modification of visibility.

Reply:
The change to activation IDs needs to be done later, it seems too fragile to fix in the same change.